### PR TITLE
Fix median to support nan

### PR DESCRIPTION
**Pull request by HU-qingqing**
## Title
Fix median operator NaN input support and add test. To fix the issue #10095
## Description
Fixed the issue where the `median` operator did not support inputs containing NaN values, and added relevant test cases to cover this scenario.

In terms of algorithm implementation, I referenced the approach used in NumPy, but with an optimized adjustment:
- For imprecise floating-point data types, I traverse the array to identify NaN values and return results immediately once NaNs are detected.
- Unlike NumPy (which handles this by adding a `k` value of `-1` in the `select` function), my implementation integrates this logic directly into the `median` function. This allows me to traverse the array early and return results upfront, eliminating the need for subsequent calls to the `select` function.

I would appreciate your feedback and review comments on whether this handling approach is appropriate.

## Type of Change
- [x] Bug fix (non-breaking change fixing an issue)
- [ ] New feature (non-breaking change adding functionality)
- [ ] Breaking change (fix or feature causing existing functionality to change)
- [x] Tests added/updated

### CODE REVIEW

**Comment by HU-qingqing** [comment]
@sklam @esc

**Comment by esc** [comment]
@HU-qingqing thank you for submitting this and helping to improve Numba. We currently do not have any capacity to review contributions like this so you are likely to experience a significant delay before we can give you any feedback. The PR has been labelled accordingly as "ready-for-review" and is now in the review queue, which has 62 PRs in it.

**Comment by HU-qingqing** [comment]
@esc Thanks for your reply! That makes total sense to me. This is indeed a fix for an edge case, and I look forward to receiving your review feedback in due course.

**Comment by esc** [comment]
@HU-qingqing in general we prefer for the Numba implementation to mirror the NumPy function in terms of expected output as far as possible (certain insignificant numerical difference may be acceptable). Asking about the added optimizations, do they only enhance the performance of the implementation or are there also corner cases where the generated output does not match what NumPy produces?

**Comment by HU-qingqing** [comment]
@esc  Thanks for your review !

**What do the optimizations do?**
When the input array contains NaNs, the optimized algorithm will return immediately after traversal, whereas NumPy will still execute the select operation in this scenario. Notably, the return value will be `NaN` — **consistent with NumPy's behavior**. This is the case this PR addresses.

No discrepancies exist in the current implementation where `overwrite_input=True` (which means the input ndarray `a` is modified in-place to save memory) is not supported.

If this parameter is to be supported in future versions, we will also need to modify the select algorithm to preserve consistent element ordering for ndarray  `a `. This is because NumPy uses the quickselect algorithm under the hood; such algorithmic differences will lead to inconsistent element ordering at non-target positions in ndarray  `a `. Moreover, since ndarray a is not part of the return value, such discrepancies are likely acceptable ?

**Comment by esc** [comment]
@HU-qingqing thank you for submitting this. In terms of test coverage, it seems like a test that exercises the code path for `supports_nans=True` is missing? Or did I miss something?

**Comment by HU-qingqing** [comment]
@esc Thank you for your prompt reply!

Any median test case with floating-point types will automatically exercise this code path. The relevant implementation is:
supports_nans = np.issubdtype(as_dtype(a.dtype), np.inexact)
and
return _median_inner(temp_arry, n, supports_nans)

The supports_nans flag is determined by the data type (dtype) rather than being a user parameter. NumPy follows the same approach since NaN values only exist for floating-point data types.

**Comment by esc** [comment]
@HU-qingqing thank you for following up.

I tried the reproducer from #10095 on this PR and discovered that this is still an issue?

```
esc@ [numba_3.14] [numba:hqq/fix_median_nan:★★₃★] ~/git/numba git show
commit ec20c640337fc65ce6852fecbc63a7ffd541787f (HEAD -> hqq/fix_median_nan)
Author: Qingqing HU <huqingqing0309@outlook.com>
Date:   Thu Dec 18 17:42:06 2025 +0800

    [FIX] format

diff --git a/numba/np/arraymath.py b/numba/np/arraymath.py
index ba4cd126df..f09db83381 100644
--- a/numba/np/arraymath.py
+++ b/numba/np/arraymath.py
@@ -1568,7 +1568,7 @@ def _select_two(arry, k, low, high):


 @register_jitable
-def _median_inner(temp_arry, n, supports_nans = False):
+def _median_inner(temp_arry, n, supports_nans=False):
     """
     The main logic of the median() call.  *temp_arry* must be disposable,
     as this function will mutate it.
@@ -1578,7 +1578,7 @@ def _median_inner(temp_arry, n, supports_nans = False):
     half = n >> 1
     if supports_nans:
         maxval = temp_arry[low]
-        for k in range(low+1, n):
+        for k in range(low + 1, n):
             if not less_than(temp_arry[k], maxval):
                 maxval = temp_arry[k]
         if np.isnan(maxval):
esc@ [numba_3.14] [numba:hqq/fix_median_nan:★★₃★] ~/git/numba ipython
Python 3.14.0 | packaged by Anaconda, Inc. | (main, Oct 22 2025, 09:01:01) [Clang 20.1.8 ]
Type 'copyright', 'credits' or 'license' for more information
IPython 9.7.0 -- An enhanced Interactive Python. Type '?' for help.
Tip: You can change the editing mode of IPython to behave more like vi, or emacs.

In [1]: import numpy as np
   ...: from numba import njit
   ...: @njit
   ...: def median_njit():
   ...:     return np.median(np.array([np.nan, 1, 2]))
   ...:
   ...: print(np.median(np.array([np.nan,1,2])))
   ...: print(median_njit())
nan
2.0
```

**Comment by esc** [comment]
I think there is an edge-case, that if the first element of the incoming array is `nan` then the algorithm fails. Using the following patch, I can get the reproducer to pass:

```
esc@ [numba_3.14] [numba:hqq/fix_median_nan:★★★₃★] ~/git/numba gid
diff --git i/numba/np/arraymath.py w/numba/np/arraymath.py
index f09db83381..4bf7be021b 100644
--- i/numba/np/arraymath.py
+++ w/numba/np/arraymath.py
@@ -1578,6 +1578,8 @@ def _median_inner(temp_arry, n, supports_nans=False):
     half = n >> 1
     if supports_nans:
         maxval = temp_arry[low]
+        if np.isnan(maxval):
+            return maxval
         for k in range(low + 1, n):
             if not less_than(temp_arry[k], maxval):
                 maxval = temp_arry[k]
```

**Comment by HU-qingqing** [comment]
@esc Yes, you're absolutely right! Great catch! I referenced NumPy's implementation logic https://github.com/numpy/numpy/blob/508943d1070d8f4e0712b98db2379d49304e4e38/numpy/_core/src/npysort/selection.cpp#L367, but actually NumPy's definition of less differs from Numba's: https://github.com/numpy/numpy/blob/508943d1070d8f4e0712b98db2379d49304e4e38/numpy/_core/src/npysort/npysort_common.h#L141, which caused the discrepancy in results. I'll fix this immediately as suggested.

**Comment by esc** [comment]
> @esc Yes, you're absolutely right! Great catch! I referenced NumPy's implementation logic https://github.com/numpy/numpy/blob/508943d1070d8f4e0712b98db2379d49304e4e38/numpy/_core/src/npysort/selection.cpp#L367, but actually NumPy's definition of less differs from Numba's: https://github.com/numpy/numpy/blob/508943d1070d8f4e0712b98db2379d49304e4e38/numpy/_core/src/npysort/npysort_common.h#L141, which caused the discrepancy in results. I'll fix this immediately as suggested.

@HU-qingqing OK, thank you. Can you add a link to the referenced implementation Logic in NumPy please? Here is an example:

https://github.com/numba/numba/blob/main/numba/np/arraymath.py#L983

Also, this may warrant a test-case for this edge case.

**Comment by esc** [comment]
For reference:

```
In [1]: def f_np(a,b):
   ...:     return (a < b or (b != b and a == a))
   ...:

In [2]: def f_nb(a,b):
   ...:     return a < b
   ...:

In [3]: import numpy as np

In [4]: for x,y in ((1, np.nan), (np.nan, 1)):
   ...:     print(f_np(x,y))
   ...:     print(f_nb(x,y))
   ...:
True
False
False
False
```

**Comment by HU-qingqing** [comment]
@esc Appreciate the review comments！

I've updated the implementation to use nan_aware_less_than instead of less_than, ensuring it matches the behavior of NumPy's less function. I've also added inline comments citing the relevant NumPy source code and supplemented the test suite with corresponding test cases.

**Comment by esc** [comment]
@HU-qingqing I took another look at this today and unfortunately it seems like the implementation suffers from performance issues

Results from `main`:

```
esc@artemis [numba_3.14] [numba:hqq/fix_median_nan:★★₄★] ~/git/numba git co main
Switched to branch 'main'
Your branch is up to date with 'origin/main'.

esc@artemis [numba_3.14] [numba:main:★★₄] ~/git/numba ipython

In [1]: from numba import njit

In [2]: import numpy as np

In [3]: a = np.arange(1e9, dtype=np.float64)

In [4]: @njit
   ...: def m(a):
   ...:     return np.median(a)
   ...:

In [5]: m(a)
Out[5]: 499999999.5

In [6]: %timeit np.median(a)
987 ms ± 41.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [7]: %timeit np.median(a)
962 ms ± 38.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [8]: %timeit np.median(a)
965 ms ± 26.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [9]: %timeit m(a)
1.09 s ± 21.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [10]: %timeit m(a)
1.09 s ± 11.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [11]: %timeit m(a)
1.1 s ± 26.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

And results from the PR:

```
esc@artemis [numba_3.14] [numba:main:★★₄] ~/git/numba git co -
Switched to branch 'hqq/fix_median_nan'

esc@artemis [numba_3.14] [numba:hqq/fix_median_nan:★★₄★] ~/git/numba ipython

In [1]: from numba import njit

In [2]: import numpy as np

In [3]: @njit
   ...: def m(a):
   ...:     return np.median(a)
   ...:

In [4]: a = np.arange(1e9, dtype=np.float64)

In [5]: m(a)
Out[5]: 499999999.5

In [6]: %timeit np.median(a)
974 ms ± 51.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [7]: %timeit np.median(a)
987 ms ± 31.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [8]: %timeit np.median(a)
971 ms ± 19.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [9]: %timeit m(a)
1.44 s ± 25.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [10]: %timeit m(a)
1.45 s ± 34.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [11]: %timeit m(a)
1.47 s ± 22.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

The linear check for `nan`s ahead of the selection is too costly.

**Comment by HU-qingqing** [comment]
@esc This optimization only applies to NaN-containing cases, and it degrades performance for most non-NaN cases. I checked NumPy's approach: it uses nan-aware less-than comparison with all pivot values stored, then does a max traversal on the latter half of the largest pivot. For median, this cuts the traversal length by at least one pass. As we don't store all pivots, my current solution is to traverse the post-half data with full NaN-aware less-than checks. This gives some performance improvement, yet the NaN check overhead still exists.

My apologies for neglecting the more general scenarios. As of now, I can think of no better workaround than reimplementing the select algorithm.

Here are my test code and test results:
```python
from numba import njit
import numpy as np
import timeit
@njit
def m(a):
   return np.median(a)
a = np.arange(1e9, dtype=np.float64)
np_op = lambda : np.median(a)
numba_op = lambda : m(a)
print(np.median(a))
print(m(a))
print(timeit.timeit(np_op, number=10) / 10)
print(timeit.timeit(numba_op, number=10) / 10)
```
```
# python3 test.py
499999999.5
499999999.5
5.042221046797931
7.066977980639786  // main 
# python3 test.py
499999999.5
499999999.5
4.964929599221795
8.424386344943196  // fix_median_nan
# python3 test.py
499999999.5
499999999.5
4.939144776016474
7.866114169608802 // local optimized version
```


local optimized version:
```
diff --git a/numba/np/arraymath.py b/numba/np/arraymath.py
index f09db8338..025e3e366 100644
--- a/numba/np/arraymath.py
+++ b/numba/np/arraymath.py
@@ -1552,16 +1552,16 @@ def _select_two(arry, k, low, high):
     """
     while True:
         assert high > low  # by construction
-        i = _partition(arry, low, high)
+        i = _partition_w_nan(arry, low, high)
         if i < k:
             low = i + 1
         elif i > k + 1:
             high = i - 1
         elif i == k:
-            _select(arry, k + 1, i + 1, high)
+            _select_w_nan(arry, k + 1, i + 1, high)
             break
         else:  # i == k + 1
-            _select(arry, k, low, i - 1)
+            _select_w_nan(arry, k, low, i - 1)
             break
 
     return arry[k], arry[k + 1]
@@ -1576,19 +1576,20 @@ def _median_inner(temp_arry, n, supports_nans=False):
     low = 0
     high = n - 1
     half = n >> 1
-    if supports_nans:
-        maxval = temp_arry[low]
-        for k in range(low + 1, n):
-            if not less_than(temp_arry[k], maxval):
-                maxval = temp_arry[k]
-        if np.isnan(maxval):
-            return maxval
-
     if n & 1 == 0:
         a, b = _select_two(temp_arry, half - 1, low, high)
-        return (a + b) / 2
+        res = (a + b) / 2
     else:
-        return _select(temp_arry, half, low, high)
+        res = _select_w_nan(temp_arry, half, low, high)
+
+    if supports_nans:
+        maxval = temp_arry[half]
+        for k in range(half + 1, n):
+            if not nan_aware_less_than(temp_arry[k], maxval):
+                maxval = temp_arry[k]
+        temp_arry[high] = maxval
+    return res
+
 
 
 @overload(np.median)
@@ -1606,7 +1607,12 @@ def np_median(a):
         n = temp_arry.shape[0]
         if not is_datetime and n == 0:
             return np.nan
-        return _median_inner(temp_arry, n, supports_nans)
+        res = _median_inner(temp_arry, n, supports_nans)
+        if supports_nans and np.isnan(temp_arry[n - 1]):
+            return np.nan
+        else:
+            return res
+
     return median_impl
```

**Comment by esc** [comment]
> My apologies for neglecting the more general scenarios. As of now, I can think of no better workaround than reimplementing the select algorithm.

No worries, good thing this was caught. I would tend to agree about having to implement the "full" select set of algorithms from NumPy. Probably quite involved. I'll bring this PR to the maintainer discussion tomorrow to learn what the other maintainers think and then get back to you.

Thank you for your work to help improve Numba, so much has been learnt.

**Comment by HU-qingqing** [comment]
@esc Hi, I have conducted additional performance tests for your reference. Since the data generated by arange is special (with lower complexity), the performance overhead introduced by adding traversal is more significant. For general cases, the performance overhead is around 8-10% (I have also tested smaller shapes). In the case of random numbers, reducing the number of traversals by half and using the nan-aware less-than operation exclusively does not yield significant benefits. Additionally, the CPU-side data differs from the previous results because I have switched to a different machine.


Here are my test code and test results:

```python
from numba import njit
import numpy as np
import timeit

@njit
def m(a):
   return np.median(a)
np_op = lambda : np.median(a)
numba_op = lambda : m(a)

time_np = 0
time_numba = 0
for i in [0, 2, 9]:
   np.random.seed(i)
   a = np.random.randn(int(1e9))
   print(np.median(a))
   print(m(a))
   time_np += timeit.timeit(np_op, number=10)
   time_numba += timeit.timeit(numba_op, number=10)

print("time numpy:", time_np / 30)
print("time numba:", time_numba / 30)

a = np.arange(1e9,dtype=np.float64)
print(np.median(a))
print(m(a))
print("time numpy arange:", timeit.timeit(np_op, number=10) / 10)
print("time numba arange:", timeit.timeit(numba_op, number=10) / 10)
```

```
python3 test.py
-1.9037215400878774e-06
-1.9037215400878774e-06
-9.728113286990556e-06
-9.728113286990556e-06
-3.334093970606604e-05
-3.334093970606604e-05
time numpy: 7.887621280550957
time numba: 10.07450645659119 // main
499999999.5
499999999.5
time numpy arange: 2.648732202872634
time numba arange: 3.196298909932375

python3 test.py
-1.9037215400878774e-06
-1.9037215400878774e-06
-9.728113286990556e-06
-9.728113286990556e-06
-3.334093970606604e-05
-3.334093970606604e-05
time numpy: 7.893340863287449
time numba: 10.785694961746534 // hqq/fix_median_nan
499999999.5
499999999.5
time numpy arange: 2.645212824642658
time numba arange: 3.9170755540952085

python3 test.py
-1.9037215400878774e-06
-1.9037215400878774e-06
-9.728113286990556e-06
-9.728113286990556e-06
-3.334093970606604e-05
-3.334093970606604e-05
time numpy: 7.890819754203161
time numba: 10.861907878890634 // local_optimized_verison
499999999.5
499999999.5
time numpy arange: 2.6632782781496642
time numba arange: 3.8066079296171664
```

**Comment by esc** [comment]
@HU-qingqing -- thank you for running those tests! It looks like the only fix that will not cause a performance hit will be to include the `nan` checking in the underlying `select` or `select_two` algorithms, the same way NumPy does this. I think as a workaround for users, a simple linear check for `nan`s can be implemented in user-code or we would recommend to use `nanmedian` for a variant of `median` that just ignores `nan`s. In case you feel inclined (and have the time and the patience) to try to tackle that, please do close this PR and open a fresh one.

**Comment by esc** [comment]
@HU-qingqing if you would like to try something more straightforward, my colleague has curated some good first issues here: https://github.com/numba/numba/issues/10408 -- in this case it's also some reductions functions that suffer from a simple lack of handling scalars.

**Comment by HU-qingqing** [comment]
@esc  Got it and fully understand this part. I'll close this PR for now and look into it further later. But do we have any standard performance tests that can be used for local or CI validation? 
Also, thanks for the follow-up information – I'm happy to contribute whatever I can.

**Comment by esc** [comment]
> @esc Got it and fully understand this part. I'll close this PR for now and look into it further later. But do we have any standard performance tests that can be used for local or CI validation?

Yes and no -- there is this:

https://github.com/numba/numba-benchmark

but it's not that well maintained and doesn't run regularly.  In general, I would argue that regular performance testing is still a very big blind spot for Numba. As such, performance regressions are not unheard of. Certainly an area that could use some effort..

> Also, thanks for the follow-up information – I'm happy to contribute whatever I can.

❤️ Awesome! The project is always in need of fresh talent. ❤️
